models/base.py

```diff
@@ -64,11 +57,12 @@ def SmartSession(input_session=None):
 
 def safe_mkdir(path):
 
+    cfg = config.Config.get()
     allowed_dirs = [
-        DATA_ROOT,
-        os.path.join(CODE_ROOT, "results"),
-        os.path.join(CODE_ROOT, "catalogs"),
-        DATA_TEMP,
+        cfg.value('path.data_root'),
+        os.path.join(_CODE_ROOT, "results"),
+        os.path.join(_CODE_ROOT, "catalogs"),
```

Comment:
Why the underscore in `_CODE_ROOT`?

Reply:
I'm thinking of it as an internal variable.

But, also: I don't like these two lines.  We shouldn't have a default data directory that's anywhere in the code directory.  I've fought with this in other packages, that want to read and write where the code lives, and it makes it really tough to use in some circumstances.  (In particular: you may well be running the code out of a place that's not writeable.)  Default data directories should all come out of the config file, and should not be hardcoded inside the code to write into the code directory.

What do you visualize as the reason for the CODE_ROOT variable?

Reply:
Actually I can see why this would be a good practice, to keep the code read-only. 
I have occasionally ran into cases where I needed the code folder, but maybe if we decide up-front that we never write to it it may not even be necessary to have that folder accessible.